File: simulator/ts/plan/shortest-dubins.ts
import { PointV, Turn } from "../dtypes"
import { Path } from "../geo/path"
import { dubins } from "./planner-dubins"

export function allDubins(loc: PointV, dest: PointV, r: number): Path[] {
  return [
    dubins(loc, dest, r, Turn.Right, Turn.Right), // rsr
    dubins(loc, dest, r, Turn.Right, Turn.Left), // rsl
    dubins(loc, dest, r, Turn.Left, Turn.Right), // lsr
    dubins(loc, dest, r, Turn.Left, Turn.Left) // lsl
  ]
    .filter((path): path is Path => !!path)
}

/**
 * Find the shortest dubins path from A to B
 */
export function shortestDubins(loc: PointV, dest: PointV, r: number): Path | undefined {
  return shortestPath(allDubins(loc, dest, r))
}

/**
 * Find the path that minimizes path length
 */
function shortestPath(paths: Path[]): Path | undefined {
  let best
  let bestScore = Infinity
  for (const path of paths) {
    const score = path.length()
    if (score < bestScore) {
      best = path
      bestScore = score
    }
  }
  return best
}

File: autopilot/lib/paralib/shortest_dubins.cpp
#include <cmath>
#include <vector>
#include "path.h"
#include "plan.h"
using namespace std;

static Path *shortest_path(vector<Path*> paths);

/**
 * Find the shortest dubins path from A to B
 */
Path *shortest_dubins(PointV loc, PointV dest, double r) {
  // Construct flight paths
  vector<Path*> paths = {
    dubins(loc, dest, r, TURN_RIGHT, TURN_RIGHT), // rsr
    dubins(loc, dest, r, TURN_RIGHT, TURN_LEFT), // rsl
    dubins(loc, dest, r, TURN_LEFT, TURN_RIGHT), // lsr
    dubins(loc, dest, r, TURN_LEFT, TURN_LEFT) // lsl
  };
  // Find the best and free the rest
  Path *shortest = shortest_path(paths);
  return shortest;
}

/**
 * Find the path that minimizes path length.
 * Free non-best plans as we go.
 */
static Path *shortest_path(vector<Path*> paths) {
  Path *best = NULL;
  double best_score = INFINITY;
  for (const auto path : paths) {
    if (path) {
      const double score = path_length(path);
      if (score < best_score) {
        if (best) {
          free_path(best);
        }
        best = path;
        best_score = score;
      } else {
        free_path(path);
      }
    }
  }
  return best;
}
